pkg/controller/util/k8s/protocol.go

```diff
@@ -0,0 +1,236 @@
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	apiPort      = 5678
+	protocolPort = 5679
+	probePort    = 5679
+)
+
+// NewClusterConfigMap returns a new ConfigMap for initializing Atomix clusters
+func NewClusterConfigMap(cluster *v1beta2.Cluster, storage *storage.RaftStorageClass, config interface{}) (*corev1.ConfigMap, error) {
+	clusterConfig, err := newNodeConfigString(cluster, storage)
+	if err != nil {
+		return nil, err
+	}
+
+	protocolConfig, err := newProtocolConfigString(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+			Labels:    cluster.Labels,
+		},
+		Data: map[string]string{
+			clusterConfigFile:  clusterConfig,
+			protocolConfigFile: protocolConfig,
+		},
+	}, nil
+}
+
+// getPodName returns the name of the pod for the given pod ID
+func getPodName(cluster *v1beta2.Cluster, pod int) string {
+	return fmt.Sprintf("%s-%d", cluster.Name, pod)
+}
+
+// getPodDNSName returns the fully qualified DNS name for the given pod ID
+func getPodDNSName(cluster *v1beta2.Cluster, pod int) string {
+	return fmt.Sprintf("%s-%d.%s.%s.svc.cluster.local", cluster.Name, pod, cluster.Name, cluster.Namespace)
+}
+
+// newNodeConfigString creates a node configuration string for the given cluster
+func newNodeConfigString(cluster *v1beta2.Cluster, storage *storage.RaftStorageClass) (string, error) {
+	members := make([]*api.MemberConfig, storage.Spec.Replicas)
+	for i := 0; i < int(storage.Spec.Replicas); i++ {
+		members[i] = &api.MemberConfig{
+			ID:           getPodName(cluster, i),
+			Host:         getPodDNSName(cluster, i),
+			ProtocolPort: protocolPort,
+			APIPort:      apiPort,
+		}
+	}
+
+	config := &api.ClusterConfig{
+		Members: members,
+	}
+
+	marshaller := jsonpb.Marshaler{}
+	return marshaller.MarshalToString(config)
+}
+
+// newProtocolConfigString creates a protocol configuration string for the given cluster and protocol
+func newProtocolConfigString(config interface{}) (string, error) {
+	bytes, err := json.Marshal(config)
+	if err != nil {
+		return "", err
+	}
+	return string(bytes), nil
+}
+
+// NewClusterDisruptionBudget returns a new pod disruption budget for the cluster group cluster
+func NewClusterDisruptionBudget(cluster *v1beta2.Cluster, storage *storage.RaftStorageClass) *policyv1beta1.PodDisruptionBudget {
+	minAvailable := intstr.FromInt(int(storage.Spec.Replicas)/2 + 1)
+	return &policyv1beta1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+		},
+		Spec: policyv1beta1.PodDisruptionBudgetSpec{
+			MinAvailable: &minAvailable,
+		},
+	}
+}
+
+// NewClusterHeadlessService returns a new headless service for a cluster group
+func NewClusterHeadlessService(cluster *v1beta2.Cluster) *corev1.Service {
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cluster.Name,
```

Comment:
The headless service name will need to have a suffix like `-hs` and the normal service should have the same name as the cluster since that's the one clients will use.